servlet-engine/src/main/java/io/micronaut/servlet/engine/DefaultServletHttpRequest.java

```diff
@@ -420,6 +428,227 @@ public void onError(Throwable t) {
         bodyContent.subscribe(s);
     }
 
+    @Override
+    public boolean isFull() {
+        return !bodyIsReadAsync;
+    }
+
+    @Override
+    public ByteBuffer<?> contents() {
+        if (bodyIsReadAsync) {
+            if (LOG.isDebugEnabled()) {
+                LOG.debug("Body is read asynchronously, cannot get contents");
+            }
+            return null;
+        }
+        try {
+            this.servletByteBuffer = new ServletByteBuffer<>(getInputStream().readAllBytes());
+            return servletByteBuffer;
+        } catch (IOException e) {
+            throw new IllegalStateException("Error getting all body contents", e);
+        }
+    }
+
+    @Override
+    public ExecutionFlow<ByteBuffer<?>> bufferContents() {
+        ByteBuffer<?> contents = contents();
+        if (contents == null) {
+            return null;
+        }
+        return ExecutionFlow.just(contents);
+    }
+
+    private final class ServletByteBuffer<T> implements ByteBuffer<T> {
```

Comment:
should not we move this class to `servlet-core` so that we are able to use it in the serverless implementations?

Reply:
Maybe extract it into its own file and annotate it with `@Experimental`

Reply:
Done in https://github.com/micronaut-projects/micronaut-servlet/commit/1bb046f1c1235034f1a5113d01a2b04d194c59e8

Which isn't showing up in the PR for unknown reasons...  github status is lying I suspect